Final updates for the cal/val delivery

gshiroma opened the pull request:
This PR adds the final updates for the cal/val delivery:
-Add mosaicking with different output formats;
-Add support to mosaicking with negative offsets;
-Add handling of Byte and Integer images; 
-Add fill value and color table to the layover shadow mask;
-Save color table and no data values into the output mosaic (when available).

Layover/shadow mask mosaic before the changes (`255` represents invalid values):
<img width="764" alt="Screenshot 2023-05-28 at 10 58 03 AM" src="https://github.com/opera-adt/RTC/assets/52007211/cfb97b80-d745-46f0-8d47-b416d67ef685">

Layover/shadow mask mosaic after the changes:
<img width="764" alt="Screenshot 2023-05-28 at 2 30 18 PM" src="https://github.com/opera-adt/RTC/assets/52007211/014bb2e1-3258-4b9a-a8bd-ac511080a7f1">

seongsujeong reviewed (commented):
Looks good overall in code wise. Most of my comments / questions are for clarity.  - Seongsu

gshiroma commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -132,6 +134,26 @@ def _compute_distance_to_burst_center(image, geotransform):
+    if no_data_value is None or np.isnan(no_data_value):
+        return np.isnan(input_array)
+
+    return input_array == no_data_value
```
Good catch! I appreciate that suggestion, @seongsujeong . Thanks!

gshiroma commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -268,7 +283,10 @@ def compute_mosaic_array(list_rtc_images, list_nlooks, mosaic_mode, scratch_dir=
 
         if verbose:
-            print(f'    mosaicking ({i+1}/{num_raster}): {os.path.basename(path_rtc)}')
+            print(f'    mosaicking ({i+1}/{num_raster}): {path_rtc}')
```
I think in this case, the fraction without spaces look better: `3/4`, `8/10` rather than `3 / 4`, `8 / 10` . What do you think?

gshiroma commented:
> Looks good overall in code wise. Most of my comments / questions are for clarity. - Seongsu

Thank you so much, @seongsujeong ! I addressed most of your comments. I just left a few minor comments related to docstrings and comments that I'll work on them tomorrow. Please, let me know if you have any other comments or concerns.

gshiroma commented:
Thank you again for your review, @seongsujeong , just finished addressing your comments. Please let me know if you have any other comments or concerns.

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -356,48 +406,82 @@ def compute_mosaic_array(list_rtc_images, list_nlooks, mosaic_mode, scratch_dir=
+                no_data_value = band_ds.GetNoDataValue()
+                gdal_dtype = band_ds.DataType
+                dtype_name = gdal.GetDataTypeName(gdal_dtype).lower()
+                if (IS_MODE_AVERAGE_ENABLED_FOR_BYTE_DTYPE is False and
```
That sounds good to me if that was the purpose of the constant. Thanks for the explanation!

seongsujeong commented on src/rtc/mosaic_geobursts.py:
```diff
@@ -132,6 +134,26 @@ def _compute_distance_to_burst_center(image, geotransform):
     return distance
 
 
+def is_invalid(input_array, no_data_value):
```
I see. Let's keep the function name then.

seongsujeong reviewed (approved):
LGTM. Tested the layover shadow mask fix, and confirmed the mask overlaid on the main product makes sense.

gshiroma commented:
Thank you, @seongsujeong !
